File: server/project-service/rabbitmq/userProducer.go
package rabbitmq

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/horatiucrisan/project-service/model"
	"github.com/horatiucrisan/project-service/utils"
	"github.com/streadway/amqp"
)

type UserProducer struct {
	conn       *amqp.Connection
	channel    *amqp.Channel
	replyQueue amqp.Queue
	queue      string
}

// NewUserProducer retrieves the queue name and generates a new rabbitMq producer
// that connects to the rabbitMq users consumer
//
// Parameters:
//   - queue: The name of the rabbitMq queue
//
// Returns:
//   - *UserProducer: The new rabbitMq producer
//   - error: An error that occured during the process
func NewUserProducer(queue string) (*UserProducer, error) {
	// Connect the producer to the rabbitmq URL
	conn, err := amqp.Dial(utils.EnvInstances.RABBITMQ_URL)
	if err != nil {
		return nil, err
	}

	// Generate a new channel
	ch, err := conn.Channel()
	if err != nil {
		return nil, err
	}

	// Generate a new rabbitMq queue
	replyQueue, err := ch.QueueDeclare(
		"",
		false,
		true,
		true,
		false,
		nil,
	)

	if err != nil {
		return nil, err
	}

	// Return the producer data
	return &UserProducer{
		conn:       conn,
		channel:    ch,
		replyQueue: replyQueue,
		queue:      queue,
	}, nil
}

// GetUsers method retrieves the user producer and sends a list of user IDs to the consumer
// and retreives the data of each user
//
// Parameters:
//   - userIds: The list of user IDs
//
// Returns:
//   - []model.User: The list of users data
//   - error: An error that occured during the process
func (p *UserProducer) GetUsers(userIds []string) ([]model.User, error) {
	// Generate a new rabbitMq queue
	replyQueue, err := p.channel.QueueDeclare(
		"",
		false,
		true,
		true,
		false,
		nil,
	)
	if err != nil {
		return nil, err
	}

	// Generate a new correlation ID of the reply queue
	correlationId := uuid.New().String()

	// Encode the users list into the JSON format
	body, err := json.Marshal(userIds)
	if err != nil {
		return nil, err
	}

	// Send the data to the rabbitMq users consumer
	err = p.channel.Publish(
		"",
		p.queue,
		false,
		false,
		amqp.Publishing{
			ContentType:   "application/json",
			CorrelationId: correlationId,
			ReplyTo:       replyQueue.Name,
			Body:          body,
		},
	)
	if err != nil {
		return nil, err
	}

	// Retreive the users data from the consumer
	msgs, err := p.channel.Consume(
		replyQueue.Name,
		"",
		true,
		true,
		false,
		false,
		nil,
	)
	if err != nil {
		return nil, err
	}

	// Timeout to reveive the data
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	for {
		select {
		// case for receiving the response message
		case msg := <-msgs:
			// check if the correlation ID is the same as the one from the producer
			if correlationId == msg.CorrelationId {
				var users []model.User
				// decode the data from the message into the users list
				err := json.Unmarshal(msg.Body, &users)
				if err != nil {
					return nil, err
				}
				return users, nil
			}
		// timeout error message for not receiving the message
		case <-ctx.Done():
			return nil, fmt.Errorf("timeout waiting for user service response")
		}
	}
}

// Close function ends the producer connection to rabbitMq
//
// Returns:
//   - error: An error that occured during the process
func (p *UserProducer) Close() error {
	// Close the channel
	if err := p.channel.Close(); err != nil {
		return err
	}

	// Close the connection
	if err := p.conn.Close(); err != nil {
		return err
	}

	return nil
}

File: server/user-lib/src/service/userProducer.ts
import amqp, {Channel, ChannelModel, connect} from "amqplib";
import env from "dotenv";
import {v4} from "uuid";
env.config();

export class UserProducer {
    private _RABBITMQ_URL: string;
    private _connection: ChannelModel | null;
    private _channel: Channel | null;

    constructor() {
        /* Verify if the env data was initialized */
        if (!process.env.RABBITMQ_URL) {
            throw new Error(`Invalid user producer env. data`);
        }

        this._RABBITMQ_URL = process.env.RABBITMQ_URL;
        this._connection = null;
        this._channel = null;
    }

    async connection(): Promise<void> {
        /* If the connection exists, exit */
        if (this._channel && this._connection) return;

        try {
            /* Initialize the connection based on the rabbitmq url */
            this._connection = await connect(this._RABBITMQ_URL);

            /* Initialize the channel using the connection */
            this._channel = await this._connection.createChannel();
        } catch (error) {
            throw new Error(`Failed to initialize user producer connection: ${error}`);
        }
    }

    /**
     * 
     * @param {string} queue The name of the queue the producer listens to
     * @param {string[]} userIds The list with the IDs of the users
     * @returns {Promise<unknown[]>} The list of users data
     */
    async assertUserQueue(queue: string, userIds: string[]): Promise<any[]> {
        /* Initialize the connection */
        await this.connection();
    
        try {
            /* If the channel of the connection failed, exit */
            if (!this._connection || !this._channel) return [];

            /* Initialize the producer to listen to the queue */
            await this._channel.assertQueue(queue, { durable: true });
    
            /* Generate a temporary queue for the response */
            const { queue: replyQueue } = await this._channel!.assertQueue('', {
                exclusive: true
            });
            
            /* Generate a correlation ID using uuid */
            const correlationId = v4();
    
            /* Return a promise that is responisble for listening to the queue,
                and replying to the temporary queue with the users data */
            return new Promise((resolve, reject) => {
                /* If the connection of channel are not initialized, exit */
                if (!this._channel || !this._connection) return [];

                /* Retrieve the message from the temporary queue */
                this._channel.consume(replyQueue, (msg) => {
                    /* Check if the correlationID of the matches matches with the one created above */
                    if (msg?.properties.correlationId === correlationId) {
                        const users = JSON.parse(msg.content.toString());

                        /* Resolve the response */
                        resolve(users); 
                    }
                }, {
                    noAck: true
                });
    
                /* Encode the list of users */
                const message = Buffer.from(JSON.stringify(userIds));

                /* Send the list of users to the users queue with the correlation ID */
                const success = this._channel?.sendToQueue(queue, message, {
                    persistent: true,
                    replyTo: replyQueue,
                    correlationId
                });
                
                /* Check if the operation succeded */
                if (!success) {
                    reject(new Error('Failed to send user IDs'));
                }
            });
    
        } catch (error) {
            await this.close();
            throw new Error(`Failed to assert user producer queue: ${error}`);
        }
    }

    async close(): Promise<void> {
        try {
            /* If the connection is closed exit */
            if (!this._channel || !this._connection) return;

            await this._channel.close();
            await this._connection.close();
        } catch (error) {
            throw new Error(`Failed to close user producer connection: ${error}`);
        }
    }
}
